From 32067d95d2b18e245991e3d5a709a64bab30f0ef Mon Sep 17 00:00:00 2001
From: inferno0230 <mail@inferno0230.in>
Date: Tue, 17 Mar 2026 01:07:15 +0530
Subject: [PATCH] display-driver: oplus_adfr: Allow force enabling of ADFR

Oplus ADFR is used by Oplus for controlling how and when LTPO works,
On Stock, Oplus have Frameworks all around the source (fwb/surfaceflinger/hwc)
and a service which does ioctl to the driver, as seen in logs

```
[  629.131598] [LCD][0][INFO][panel_ioctl:368] pid = 1793, cmd = PANEL_IOCTL_GET_DYNAMIC_TE
[  629.284651] [LCD][0][INFO][panel_ioctl:368] pid = 1793, cmd = PANEL_IOCTL_GET_DYNAMIC_TE
[  629.438330] [LCD][0][INFO][panel_ioctl:368] pid = 1793, cmd = PANEL_IOCTL_GET_DYNAMIC_TE
[  629.592818] [LCD][0][INFO][panel_ioctl:368] pid = 1793, cmd = PANEL_IOCTL_GET_DYNAMIC_TE
[  629.746732] [LCD][0][INFO][panel_ioctl:368] pid = 1793, cmd = PANEL_IOCTL_GET_DYNAMIC_TE
[  629.900161] [LCD][0][INFO][panel_ioctl:368] pid = 1793, cmd = PANEL_IOCTL_GET_DYNAMIC_TE
[  630.052446] [LCD][0][INFO][panel_ioctl:368] pid = 1793, cmd = PANEL_IOCTL_GET_DYNAMIC_TE
[  630.205830] [LCD][0][INFO][panel_ioctl:368] pid = 1793, cmd = PANEL_IOCTL_GET_DYNAMIC_TE
[  630.362664] [LCD][0][INFO][panel_ioctl:368] pid = 1793, cmd = PANEL_IOCTL_GET_DYNAMIC_TE
[  630.516080] [LCD][0][INFO][panel_ioctl:368] pid = 1793, cmd = PANEL_IOCTL_GET_DYNAMIC_TE
[  630.670454] [LCD][0][INFO][panel_ioctl:368] pid = 1793, cmd = PANEL_IOCTL_GET_DYNAMIC_TE
```
This service is vendor.oplus.hardware.displaypanelfeature

Since we don't have all those things, Add a kernel node and
a kernel sa fallback to enable auto adfr commands.

Our HWC/SF still won't be aware of the refresh rate changes since the
timings aren't actually exposed. Actual refresh rates can be overserved
by reading values from /sys/kernel/oplus_display/test_te.

Change-Id: I31f7170c2560aaa17ca1aa2b8c77f5e61b54c316
Signed-off-by: inferno0230 <mail@inferno0230.in>
---
 .../display-drivers/oplus/oplus_adfr.c        | 72 +++++++++++++++++++
 .../display-drivers/oplus/oplus_adfr.h        |  6 ++
 .../oplus/oplus_display_private_api.c         |  2 +
 3 files changed, 80 insertions(+)

diff --git a/qcom/opensource/display-drivers/oplus/oplus_adfr.c b/qcom/opensource/display-drivers/oplus/oplus_adfr.c
index 32cb48fe..e38533b1 100644
--- a/qcom/opensource/display-drivers/oplus/oplus_adfr.c
+++ b/qcom/opensource/display-drivers/oplus/oplus_adfr.c
@@ -75,6 +75,7 @@ unsigned int oplus_adfr_display_id = OPLUS_ADFR_PRIMARY_DISPLAY;
 EXPORT_SYMBOL(oplus_adfr_display_id);
 /* adfr global structure */
 static struct oplus_adfr_params g_oplus_adfr_params[2] = {0};
+static bool oplus_adfr_ltpo_enabled = false;
 
 /* -------------------- extern -------------------- */
 /* extern params */
@@ -1516,6 +1517,26 @@ static int oplus_adfr_min_fps_check(void *dsi_panel, unsigned int min_fps)
 	return min_fps;
 }
 
+static int oplus_adfr_sa_fallback_min_fps_get(void *dsi_panel)
+{
+	struct dsi_panel *panel = dsi_panel;
+	struct dsi_display_mode_priv_info *priv_info = NULL;
+
+	if (!panel || !panel->cur_mode || !panel->cur_mode->priv_info) {
+		ADFR_ERR("invalid panel params\n");
+		return 0;
+	}
+
+	priv_info = panel->cur_mode->priv_info;
+	if (!priv_info->oplus_adfr_min_fps_mapping_table_count) {
+		ADFR_DEBUG("no sa fallback because min fps mapping table is not set\n");
+		return 0;
+	}
+
+	return priv_info->oplus_adfr_min_fps_mapping_table[
+		priv_info->oplus_adfr_min_fps_mapping_table_count - 1];
+}
+
 static int oplus_adfr_min_fps_update(void *dsi_display, unsigned int min_fps)
 {
 	int rc = 0;
@@ -1686,6 +1707,7 @@ int oplus_adfr_sa_handle(void *sde_encoder_virt)
 {
 	int rc = 0;
 	unsigned int h_skew = STANDARD_ADFR;
+	unsigned int fallback_min_fps = 0;
 	struct sde_encoder_virt *sde_enc = sde_encoder_virt;
 	struct sde_connector *c_conn = NULL;
 	struct dsi_display *display = NULL;
@@ -1761,6 +1783,21 @@ int oplus_adfr_sa_handle(void *sde_encoder_virt)
 
 	OPLUS_ADFR_TRACE_BEGIN("oplus_adfr_sa_handle");
 
+	if (oplus_adfr_ltpo_enabled && !p_oplus_adfr_params->sa_fallback_applied
+			&& !p_oplus_adfr_params->need_filter_auto_on_cmd) {
+		fallback_min_fps = oplus_adfr_sa_fallback_min_fps_get(display->panel);
+		if (fallback_min_fps
+				&& (fallback_min_fps < display->panel->cur_mode->timing.refresh_rate)) {
+			p_oplus_adfr_params->auto_mode = OPLUS_ADFR_AUTO_ON;
+			p_oplus_adfr_params->auto_mode_updated = true;
+			p_oplus_adfr_params->sa_min_fps = fallback_min_fps;
+			p_oplus_adfr_params->sa_min_fps_updated = true;
+			p_oplus_adfr_params->sa_fallback_applied = true;
+			ADFR_INFO("enable kernel sa fallback, auto_mode:%u, sa_min_fps:%u\n",
+					p_oplus_adfr_params->auto_mode, p_oplus_adfr_params->sa_min_fps);
+		}
+	}
+
 	if (p_oplus_adfr_params->auto_mode_updated) {
 		rc = oplus_adfr_auto_mode_update(display, p_oplus_adfr_params->auto_mode);
 		if (rc) {
@@ -1862,6 +1899,7 @@ int oplus_adfr_status_reset(void *dsi_panel)
 
 	h_skew = panel->cur_mode->timing.h_skew;
 	refresh_rate = panel->cur_mode->timing.refresh_rate;
+	p_oplus_adfr_params->sa_fallback_applied = false;
 
 	if ((h_skew == STANDARD_ADFR) || (h_skew == STANDARD_MFR)) {
 		p_oplus_adfr_params->auto_mode = OPLUS_ADFR_AUTO_OFF;
@@ -5948,6 +5986,40 @@ ssize_t oplus_adfr_get_test_te_attr(struct kobject *obj,
 
 	return sprintf(buf, "%u\n", refresh_rate);
 }
+ssize_t oplus_adfr_set_ltpo_enable_attr(struct kobject *obj,
+	struct kobj_attribute *attr, const char *buf, size_t count)
+{
+	unsigned int ltpo_enabled = 0;
+
+	ADFR_DEBUG("start\n");
+
+	if (!buf) {
+		ADFR_ERR("invalid buf param\n");
+		return count;
+	}
+
+	if (sscanf(buf, "%u", &ltpo_enabled) != 1) {
+		ADFR_ERR("failed to parse ltpo value\n");
+		return count;
+	}
+
+	oplus_adfr_ltpo_enabled = !!ltpo_enabled;
+	ADFR_INFO("ltpo_enable:%u\n", oplus_adfr_ltpo_enabled);
+
+	ADFR_DEBUG("end\n");
+
+	return count;
+}
+ssize_t oplus_adfr_get_ltpo_enable_attr(struct kobject *obj,
+	struct kobj_attribute *attr, char *buf)
+{
+	if (!buf) {
+		ADFR_ERR("invalid buf param\n");
+		return -EINVAL;
+	}
+
+	return sprintf(buf, "%u\n", oplus_adfr_ltpo_enabled ? 1 : 0);
+}
 ssize_t oplus_display_set_high_precision_rscc(struct kobject *obj,
 		struct kobj_attribute *attr,
 		const char *buf, size_t count)
diff --git a/qcom/opensource/display-drivers/oplus/oplus_adfr.h b/qcom/opensource/display-drivers/oplus/oplus_adfr.h
index 04ba2660..d33946ec 100644
--- a/qcom/opensource/display-drivers/oplus/oplus_adfr.h
+++ b/qcom/opensource/display-drivers/oplus/oplus_adfr.h
@@ -148,6 +148,7 @@ struct oplus_adfr_params {
 	bool need_filter_auto_on_cmd;					/* indicates whether auto on cmds need to be filtered if auto off cmds have been sent within one frame or not */
 	unsigned int sa_min_fps;						/* the minimum self-refresh rate when no image would be sent to ddic in sa mode */
 	bool sa_min_fps_updated;						/* indicates whether sa min fps is updated or not */
+	bool sa_fallback_applied;						/* indicates whether kernel fallback has armed sa auto mode after the latest mode reset */
 	bool skip_min_fps_setting;						/* indicates whether min fps setting should be skipped or not */
 	unsigned int sw_fps;							/* software vsync value */
 	unsigned int fakeframe;							/* indicates whether fakeframe is enabled or not */
@@ -329,6 +330,11 @@ ssize_t oplus_adfr_set_test_te_attr(struct kobject *obj,
 	struct kobj_attribute *attr, const char *buf, size_t count);
 ssize_t oplus_adfr_get_test_te_attr(struct kobject *obj,
 	struct kobj_attribute *attr, char *buf);
+/* ltpo enable */
+ssize_t oplus_adfr_set_ltpo_enable_attr(struct kobject *obj,
+	struct kobj_attribute *attr, const char *buf, size_t count);
+ssize_t oplus_adfr_get_ltpo_enable_attr(struct kobject *obj,
+	struct kobj_attribute *attr, char *buf);
 /* high precision test */
 ssize_t oplus_display_set_high_precision_rscc(struct kobject *obj,
 	struct kobj_attribute *attr, const char *buf, size_t count);
diff --git a/qcom/opensource/display-drivers/oplus/oplus_display_private_api.c b/qcom/opensource/display-drivers/oplus/oplus_display_private_api.c
index a2409bf4..d6ebc35a 100644
--- a/qcom/opensource/display-drivers/oplus/oplus_display_private_api.c
+++ b/qcom/opensource/display-drivers/oplus/oplus_display_private_api.c
@@ -3261,6 +3261,7 @@ static OPLUS_ATTR(ffc_mode, S_IRUGO|S_IWUSR, oplus_get_ffc_mode_debug,
 static OPLUS_ATTR(adfr_config, S_IRUGO | S_IWUSR, oplus_adfr_get_config_attr, oplus_adfr_set_config_attr);
 static OPLUS_ATTR(mux_vsync_switch, S_IRUGO | S_IWUSR, oplus_adfr_get_mux_vsync_switch_attr, oplus_adfr_set_mux_vsync_switch_attr);
 static OPLUS_ATTR(test_te, S_IRUGO | S_IWUSR, oplus_adfr_get_test_te_attr, oplus_adfr_set_test_te_attr);
+static OPLUS_ATTR(ltpo_enable, S_IRUGO | S_IWUSR, oplus_adfr_get_ltpo_enable_attr, oplus_adfr_set_ltpo_enable_attr);
 #endif /* OPLUS_FEATURE_DISPLAY_ADFR */
 #ifdef OPLUS_FEATURE_DISPLAY_HIGH_PRECISION
 static OPLUS_ATTR(high_precision_rscc_set, S_IRUGO | S_IWUSR, oplus_display_get_high_precision_rscc,
@@ -3331,6 +3332,7 @@ static struct attribute *oplus_display_attrs[] = {
 	&oplus_attr_adfr_config.attr,
 	&oplus_attr_mux_vsync_switch.attr,
 	&oplus_attr_test_te.attr,
+	&oplus_attr_ltpo_enable.attr,
 #endif /* OPLUS_FEATURE_DISPLAY_ADFR */
 #ifdef OPLUS_FEATURE_DISPLAY_HIGH_PRECISION
 	&oplus_attr_high_precision_rscc_set.attr,
-- 
2.43.0

